### PR TITLE
[stdlib] Comprehensive array tests

### PR DESCRIPTION
**Pull request by lancep**
This brings the Array unit tests up to par with the Dictionary ones

rdar://problem/18272070

**Test timings**
before:
real	0m13.050s
user	0m12.216s
sys	0m0.959s

after: 
real	0m16.882s
user	0m15.828s
sys	0m1.217s

### CODE REVIEW

**Comment by lancep** [comment]
@swift-ci please smoke test

**Comment by lancep** [comment]
@swift-ci please smoke test Linux platform

**Comment by lancep** [comment]
@swift-ci please smoke test Linux platform

**Comment by lancep** [comment]
@swift-ci please smoke test Linux platform

**Review by milseman — APPROVED** [review]
LGTM

**Comment by lancep** [comment]
@swift-ci please smoke test Linux platform

File: validation-test/stdlib/Arrays.swift.gyb [hunk 1]
```diff
@@ -97,6 +113,1393 @@ ArrayTestSuite.test("valueDestruction") {
...
+    expectEqual(1010, a1[0].value)
+
+    var a2 = a1
+    a2.removeAll()
+    let identity2 = a2._rawIdentifier()
+    expectEqual(identity1, a1._rawIdentifier())
+    expectNotEqual(identity2, identity1)
+    expectEqual(3, a1.count)
+    expectEqual(1010, a1[0].value)
+    assert(a2._buffer.capacity < originalCapacity)
```
Comment (moiseev) [inline validation-test/stdlib/Arrays.swift.gyb]: Asserts are not useful in tests. If there is a condition, better check it with one of the `expect` functions.

File: validation-test/stdlib/Arrays.swift.gyb [hunk 2]
```diff
@@ -97,6 +113,1393 @@ ArrayTestSuite.test("valueDestruction") {
...
+  let identity1 = a._rawIdentifier()
+  expectTrue(isCocoaArray(a))
+
+  var iter = a.makeIterator()
+  var values = Array<Int>()
+  while let value = iter.next() {
+    values.append((value as! TestObjCValueTy).value)
+  }
+  expectEqual(values, [1010, 1020, 1030])
+  // The following is not required by the IteratorProtocol protocol, but
```
Comment (moiseev) [inline validation-test/stdlib/Arrays.swift.gyb]: Actually... This comment is no longer true. See: https://github.com/apple/swift-evolution/blob/master/proposals/0052-iterator-post-nil-guarantee.md

File: validation-test/stdlib/Arrays.swift.gyb [hunk 3]
```diff
@@ -97,6 +113,1393 @@ ArrayTestSuite.test("valueDestruction") {
...
+  expectEqual(3, CustomImmutableNSArray.timesObjectAtIndexWasCalled)
+  expectNotEqual(0, CustomImmutableNSArray.timesCountWasCalled)
+  expectEqual(3, TestBridgedValueTy.bridgeOperations)
+
+  let bridgedBack: NSArray = convertArrayToNSArray(a)
+  expectNotEqual(
+    unsafeBitCast(nsa, to: Int.self),
+    unsafeBitCast(bridgedBack, to: Int.self))
+
+  _fixLifetime(nsa)
```
[the remark this answers is not in the document]
Comment (lancep) [reply validation-test/stdlib/Arrays.swift.gyb]: I did not, meant to use the latter here.

**Comment by lancep** [comment]
@swift-ci please smoke test

**Comment by lancep** [comment]
@swift-ci please smoke test

**Comment by lancep** [comment]
@swift-ci please clean smoke test Linux platform

**Pull request merged**